rasa/core/trackers.py

```diff
@@ -384,30 +385,48 @@ def _undo_till_previous(event_type: Type[Event], done_events: List[Event]) -> No
                 break
 
     @staticmethod
-    def _first_loop_execution_or_previous_rejection(
+    def _first_loop_execution_or_unhappy_path(
         loop_action_name: Text, applied_events: List[Event]
     ) -> bool:
+        next_action_in_the_future: Optional[Text] = None
+
         for event in reversed(applied_events):
-            # Form was rejected previously.
-            if (
-                isinstance(event, ActionExecutionRejected)
-                and event.action_name == loop_action_name
-            ):
+            form_was_not_active = isinstance(event, Form) and event.name is None
+            if form_was_not_active:
                 return True
 
-            # Form might have run before but was deactivated and is now running again.
-            if isinstance(event, Form) and event.name is None:
+            if DialogueStateTracker._is_unhappy_path(
+                loop_action_name, event, next_action_in_the_future
+            ):
                 return True
 
-            # Form ran before.
-            if (
-                isinstance(event, ActionExecuted)
-                and event.action_name == loop_action_name
-            ):
+            if not isinstance(event, ActionExecuted):
+                continue
+
+            form_ran_before = event.action_name == loop_action_name
+            if form_ran_before:
                 return False
 
+            next_action_in_the_future = event.action_name
+
         return True
 
+    @staticmethod
+    def _is_unhappy_path(
```

Comment:
I think it might be correct, but the logic is confusing. The name suggests that you check the whole path, while this method checks event by event

Reply:
`might be correct` :-D very convincing 😆